csp/tests/adapters/test_websocket.py

```diff
@@ -13,28 +14,70 @@
     import tornado.web
     import tornado.websocket
 
-    from csp.adapters.websocket import JSONTextMessageMapper, RawTextMessageMapper, Status, WebsocketAdapterManager
+    from csp.adapters.websocket import (
+        ActionType,
+        ConnectionRequest,
+        JSONTextMessageMapper,
+        RawTextMessageMapper,
+        Status,
+        WebsocketAdapterManager,
+        WebsocketHeaderUpdate,
+        WebsocketStatus,
+    )
 
     class EchoWebsocketHandler(tornado.websocket.WebSocketHandler):
         def on_message(self, msg):
+            # Carve-out to allow inspecting the headers
+            if msg == "header1":
+                msg = self.request.headers.get(msg, "")
             return self.write_message(msg)
 
-
-@unittest.skipIf(not os.environ.get("CSP_TEST_WEBSOCKET"), "Skipping websocket adapter tests")
-class TestWebsocket(unittest.TestCase):
-    @classmethod
-    def setUpClass(cls):
-        cls.app = tornado.web.Application([(r"/", EchoWebsocketHandler)])
-        cls.app.listen(8000)
-        cls.io_loop = tornado.ioloop.IOLoop.current()
-        cls.io_thread = threading.Thread(target=cls.io_loop.start)
-        cls.io_thread.start()
-
-    @classmethod
-    def tearDownClass(cls):
-        cls.io_loop.add_callback(cls.io_loop.stop)
-        if cls.io_thread:
-            cls.io_thread.join()
+    @contextmanager
+    def create_tornado_server(port: int):
+        """Base context manager for creating a Tornado server in a thread"""
+        ready_event = threading.Event()
+        io_loop = None
+        app = None
+        io_thread = None
+
+        def run_io_loop():
+            nonlocal io_loop, app
+            io_loop = tornado.ioloop.IOLoop()
+            io_loop.make_current()
+            app = tornado.web.Application([(r"/", EchoWebsocketHandler)])
+            app.listen(port)
+            ready_event.set()
+            io_loop.start()
+
+        io_thread = threading.Thread(target=run_io_loop)
+        io_thread.start()
+        ready_event.wait()
+
+        try:
+            yield io_loop, app, io_thread
+        finally:
+            io_loop.add_callback(io_loop.stop)
+            if io_thread:
+                io_thread.join(timeout=5)
+                if io_thread.is_alive():
+                    raise RuntimeError("IOLoop failed to stop")
+
+    @contextmanager
+    def tornado_server(port: int = 8001):
+        """Simplified context manager that uses the base implementation"""
+        with create_tornado_server(port) as (_io_loop, _app, _io_thread):
+            yield
+
+
+@pytest.mark.skipif(os.environ.get("CSP_TEST_WEBSOCKET") is None, reason="'CSP_TEST_WEBSOCKET' env variable is not set")
```

Comment:
if these don't rely on an external service like the kafka adapter does, it can be included

Reply:
They don't, they use tornado to run a local server on port 8000/1

Reply:
Getting a weird error on macos-14 (which is what I use locally and it works :/ ) which seems to be unrelated to these changes since it occurs by just running the tests on CICD without these changes, see here: https://github.com/Point72/csp/actions/runs/12026657375/job/33526065312?pr=397

Reply:
Also, not sure how to avoid address in use errors, like if multiple of the tests are running at the same time